overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,578 @@
+	return m.prompts.PromptWithID(userID, promptID)
+}
+
+// HandleReply checks that the given reply contents are valid, satisfies the
+// original request, and does not conflict with any existing rules (if lifespan
+// is not "single"). If all of these are true, sends a reply for the prompt with
+// the given ID, and both creates a new rule and checks any outstanding prompts
+// against it, if the lifespan is not "single".
+func (m *InterfacesRequestsManager) HandleReply(userID uint32, promptID prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (satisfiedPromptIDs []prompting.IDType, retErr error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	prompt, err := m.prompts.PromptWithID(userID, promptID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Outcome and lifesnap are validated while unmarshalling, and duration is
+	// validated when the rule is being added. So only need to validate
+	// constraints.
+	if err := constraints.ValidateForInterface(prompt.Interface); err != nil {
+		return nil, err
+	}
+
+	// Check that constraints matches original requested path.
+	// AppArmor is responsible for pre-vetting that all paths which appear
+	// in requests from the kernel are allowed by the appropriate
+	// interfaces, so we do not assert anything else particular about the
+	// constraints, such as check that the path pattern does not match
+	// any paths not granted by the interface.
+	// TODO: Should this be reconsidered?
+	matches, err := constraints.Match(prompt.Constraints.Path())
+	if err != nil {
+		return nil, err
+	}
+	if !matches {
+		return nil, fmt.Errorf("constraints in reply do not match original request: '%v' does not match '%v'; please try again", constraints, prompt.Constraints)
+	}
+
+	// XXX: do we want to allow only replying to a select subset of permissions, and
+	// auto-deny the rest?
+	contained := constraints.ContainPermissions(prompt.Constraints.RemainingPermissions())
+	if !contained {
+		return nil, fmt.Errorf("replied permissions do not include all requested permissions: requested %v, replied %v; please try again", prompt.Constraints.RemainingPermissions(), constraints.Permissions)
+	}
+
+	// It is important that a lock is held while checking for conflicts with
+	// other rules so that if the rule is eventually removed due to an error,
+	// no prompts can have been matched against it in the meantime.
+	var newRule *requestrules.Rule
+	if lifespan != prompting.LifespanSingle {
+		// Check that adding the rule doesn't conflict with other rules
+		newRule, err = m.rules.AddRule(userID, prompt.Snap, prompt.Interface, constraints, outcome, lifespan, duration)
+		if err != nil {
+			// Rule conflicts with existing rule (at least one identical pattern
+			// variant and permission). This should be considered a bad reply,
+			// since the user should only be prompted for permissions and paths
+			// which are not already covered.
+			return nil, err
```

Comment:
we always end up here in tests according to coverage, what am I missing? that we don't support other Lifespans in rules atm?

Reply:
We don't support rules with `LifespanSingle` (hence the checks when we're replying to a prompt, for which a reply can have `LifespanSingle`), but the tests should add `LifespanForever` and `LifespanTimespan` rules. I think all should be exercised now?